doc/architectural/goto-program-transformations.md

```diff
@@ -0,0 +1,249 @@
+\ingroup module_hidden
+
+\page goto-program-transformations Goto Program Transformations
+
+TODO:
+- Add link for each pass to the function that implements it.
+- Add to each title the name of the function that implements it.
+- Add link points and cross-linking
+- Fix consistency of Removal/Remove etc.
+
+## Core Transformation Passes
+
+This section lists the transformation passes that must have been applied for the
+goto model to be in core goto format.
+
+Note that the passes are listed below in the order they are currently called in
+CBMC. While all dependencies on the ordering are not fully documented, the
+following order should be used.
+
+### Removal/Lowering of Assembly
+
+This pass goes through the goto model and removes or lowers instances of
+assembly intructions. Assembly instructions are stored in instances of the
+`other` instruction.
+
+_This pass has no predeccesor._
+
+### String Instrumentation
+
+This (optional) pass adds checks on calls to some of the C standard library
+string functions. It uses the tracking information added by the
+[string-abstraction] pass. It is switched on by the `--string-abstraction`
+command line option.
+
+_Predecessor passes are [Removal/Lowering of Assembly] or
+[Linking of Standard Libraries]._
+
+### Removal/Lowering of Function Pointers
+
+This pass finds all the instructions which are function calls to the value of a
+function pointer. Each instruction is then replaced with a switch block. The
+switch block contains a case for each of the potential targets of the  function
+pointer. The targets are found by looking for all functions in the goto program
+that match the type of the function pointer.
+
+Note that this pass must be applied after all linking has been done. Either for
+the standard libraries [link] or when combining multiple goto models into one.
+
+_Predecessor pass is [String Instrumentation]._
+
+
+### Memory Mapped IO Instrumentation
+
+This pass finds pointer dereferences and adds corresponding calls to the
+`__CPROVER_mm_io_r` and `__CPROVER_mm_io_w` modelling functions if they exist.
+See [modeling-mmio.md] for details of modeling memory-mapped I/O regions of
+device interfaces. This pass is always carried out but will only make changes if
+one of the modelling functions exist.
+
+_Predecessor pass is [Removal/Lowering of Function Pointers]._
+
+
+### Instrument/Remove Preconditions
+
+This pass moves preconditions associated with functions to call sites rather
```

Comment:
I am not sure. @tautschnig I think you may be more familiar with the contracts work than me, though I think this pass may intended for models of library functions, rather than contracts - Can you share any insights into why we need to do this with preconditions, but not postconditions?